### PR TITLE
Get the IID token before fetching Remote Config.

### PR DESCRIPTION
**Pull request by danasilver**
Remote Config requires an IID token to evaluate certain conditions, but does not currently wait for the token from `FirebaseInstanceId#getToken()` before fetching. This returns the default condition value in some cases (ex. percentage conditions) rather than the evaluated result for an app instance.

This change replaces the use of the deprecated `#getToken()` with `#getInstanceId()`, which returns an async `Task` we can wait to complete before fetching, without otherwise blocking.

### CODE REVIEW

**Comment by clp93** [comment]
Does this change slow down first fetch? Would adding a call to `getInstanceId()` in `FirebaseRemoteConfig#ensureInitialized()` or somewhere else in `FirebaseRemoteConfig` be appropriate?

**Comment by danasilver** [comment]
It would slow down first fetch in the case where the token is not yet available (first fetch after installing). Agreed that adding a call to `FirebaseRemoteConfig#ensureInitialized()` seems appropriate since the IID token is required for fetch now.

We discussed possibly kicking off a call to `getInstanceId()` when the `FirebaseApp` is initialized, but I'm not sure that's feasible purely within the Remote Config SDK.

File: firebase-config/src/test/java/com/google/firebase/remoteconfig/FirebaseRemoteConfigTest.java [hunk 1]
```diff
@@ -196,6 +205,7 @@ public void setUp() throws Exception {
   public void ensureInitialized_notInitialized_isNotComplete() {
     loadCacheWithConfig(mockFetchedCache, /*container=*/ null);
     loadCacheWithConfig(mockDefaultsCache, /*container=*/ null);
+    when(mockFirebaseInstanceId.getInstanceId()).thenReturn(Tasks.forResult(INSTANCE_ID_RESULT));
```
Comment (clp93) [inline firebase-config/src/test/java/com/google/firebase/remoteconfig/FirebaseRemoteConfigTest.java]: nit: preserve order?

Comment (clp93) [inline firebase-config/src/test/java/com/google/firebase/remoteconfig/FirebaseRemoteConfigTest.java]: in previous android changes tests, we have used methods like `loadInstanceIdAndToken()` (which you changed below). WDYT about doing something similar in this file? It does make it slightly more readable, but at this point, i'm just infavor of consistency :)  https://osscs.corp.google.com/firebase/firebase-android-sdk/+/master:firebase-config/src/test/java/com/google/firebase/remoteconfig/internal/ConfigFetchHandlerTest.java;l=868-871;drc=5301121df4727b357293b3b7de00555362f90170

File: firebase-config/src/main/java/com/google/firebase/remoteconfig/internal/ConfigFetchHandler.java [hunk 2]
```diff
@@ -270,15 +286,16 @@ private String createThrottledMessage(long throttledDurationInMillis) {
...
+  private FetchResponse fetchFromBackend(InstanceIdResult instanceId, Date currentTime)
+      throws FirebaseRemoteConfigException {
     try {
       HttpURLConnection urlConnection = frcBackendApiClient.createHttpURLConnection();
 
       FetchResponse response =
           frcBackendApiClient.fetch(
               urlConnection,
-              firebaseInstanceId.getId(),
-              firebaseInstanceId.getToken(),
```
Comment (erikeldridge) [inline firebase-config/src/main/java/com/google/firebase/remoteconfig/internal/ConfigFetchHandler.java]: Thinking out loud: this continues to be an instance var. Weird how it has `getInstanceId` and `getId` methods.

File: firebase-config/src/test/java/com/google/firebase/remoteconfig/FirebaseRemoteConfigTest.java (hunk 1 above)
Comment (danasilver) [reply firebase-config/src/test/java/com/google/firebase/remoteconfig/FirebaseRemoteConfigTest.java]: I was ordering by complete then incomplete tasks, but I can use the order of tasks in `ensureInitialized` instead

Comment (danasilver) [reply firebase-config/src/test/java/com/google/firebase/remoteconfig/FirebaseRemoteConfigTest.java]: Sure, I like the consistency. I do think just having the mock in the method is easier to read, though.

File: firebase-config/src/main/java/com/google/firebase/remoteconfig/internal/ConfigFetchHandler.java [hunk 3]
```diff
@@ -188,7 +189,21 @@ public ConfigFetchHandler(
                   createThrottledMessage(backoffEndTime.getTime() - currentTime.getTime()),
                   backoffEndTime.getTime()));
     } else {
-      fetchResponseTask = fetchFromBackendAndCacheResponse(currentTime);
+      Task<InstanceIdResult> instanceIdTask = firebaseInstanceId.getInstanceId();
+      fetchResponseTask =
+          instanceIdTask.continueWithTask(
+              executor,
+              (completedIidTask) -> {
```
[the remark this answers is not in the document]
Comment (erikeldridge) [reply firebase-config/src/main/java/com/google/firebase/remoteconfig/internal/ConfigFetchHandler.java]: 👍

File: firebase-config/src/main/java/com/google/firebase/remoteconfig/internal/ConfigFetchHandler.java (hunk 2 above)
Comment (erikeldridge) [reply firebase-config/src/main/java/com/google/firebase/remoteconfig/internal/ConfigFetchHandler.java]: I was just noting the similarity in naming between `firebaseInstanceId.getInstanceId` and `instanceId.getId`. I guess this is highlighted by this review, where we can see both side by side.  To be clear, I don't think any action is required.

File: firebase-config/src/main/java/com/google/firebase/remoteconfig/internal/ConfigFetchHandler.java [hunk 4]
```diff
@@ -188,7 +189,21 @@ public ConfigFetchHandler(
...
+                if (!completedIidTask.isSuccessful()) {
+                  return Tasks.forException(
+                      new FirebaseRemoteConfigClientException(
+                          "Failed to get Firebase Instance ID token for fetch.",
+                          instanceIdTask.getException()));
```
Comment (ciarand) [inline firebase-config/src/main/java/com/google/firebase/remoteconfig/internal/ConfigFetchHandler.java]: nit: just grab the exception from the callback arg

**Pull request merged**